### DISCUSSION TITLE
Does Nitro have support for middleware?

### DISCUSSION THREAD

**Discussion by OptimusePrime** [nitrojs/nitro#338] [Q&A] [answered]
Does Nitro have support for middleware: something similar to Express middleware? If so, how do you define one?

**Comment by pi0** [comment, marked as answer]
Hi. Yes it does. You have to simply put them into `server/middleware` directory 🪄 

https://v3.nuxtjs.org/guide/features/server-routes#server-middleware

**Reply by OptimusePrime** [reply]
Woah, thanks for such a quick answer! I figured it out on my own, but I used `defineMiddleware` from `h3` instead. Is there a way to scope middleware to a specific route and is there a way to access headers from an event handler (edit: never mind, I figured out the headers)? For anyone else: `event.req.headers` and then just use the dot notation to select the header name, e.g. `event.req.headers.apiKey`.

**Reply by Baloche** [reply]
Also interested in a way to scope middleware to specific routes 🙂 . My use case would be a middleware that return the logged in user from a jwt in the req headers or throw a 401 if the jwt is not valid or the user doesnt exist, but only for protected routes.

**Reply by Shahriar-Rahi** [reply]
@OptimusePrime @Baloche did any of you get any way to scope middleware to specific routes?

**Reply by pi0** [reply]
You can simply use an `if` condition to skip middleware and only run it for selected routes:

```js
export default eventHandler((event) => {
  if (!event.req.url.startsWith('/api/auth')) {
    return;
  }
  return 'Not authenticated';
});
```

https://stackblitz.com/edit/github-9wwzqi?file=server%2Fmiddleware%2Ftest.ts

But you can also instead of (classic) middleware, opt to use composable utilities in order to check auth on routes you need. I highly recommend this approach and it is what h3 utils are designed for and how are made to work.

Check this example for using this approach: https://codesandbox.io/s/nuxt3-iron-session-bw3z4m?file=/server/utils/session.ts

**Reply by Shahriar-Rahi** [reply]
Thanks for the reply @pi0. Skipping middleware using `if` condition won't work in my case. I was trying to create API for my Nuxt3 project following h3 documentation. Had to use next in my case, but next() is not shifting control to the next handler as it should in specific routes, hence it is not executing the next middleware in the middleware stack. next() only works if we apply middleware to all routes like `app.use((req, res, next) => next())`. Here's what I tried to do to create nested router:
```
import { createApp, createRouter, useBase } from 'h3';

const app = createApp();
const router = createRouter();

router.get('/v1/test', (req, res, next) => { next(); }, (req, res, next) => { res.end(); });
router.get('/v1/hello', (req, res) => { res.end('hello'); });

app.use(router);
export default useBase('/api', app);
```
getting 404 for `/api/v1/test`, other routes like `/api/v1/hello` working fine.
Am I doing something wrong or is there any solution to this problem?

**Reply by pi0** [reply]
Have you tried to instead of `next()` to use `return;`;? H3 moves to the next middleware by doing this.

**Reply by Shahriar-Rahi** [reply]
Tried without next(), still getting 404

**Reply by pi0** [reply]
Oh i see! You are creating a sub-app! It is not a middleware anymore. 404 you are getting is probably from (your) app stack not Nitro. Any reason not using `/api` directory to make this two routes?

**Reply by Shahriar-Rahi** [reply]
actually I'm using the `/api` directory, the code snippet I gave is in `server/api/v1/[...].js` file.

**Reply by pi0** [reply]
I understand that you probably prefer to use single file and that we might need to fix it. But normally, having `/api/v1/test.ts` and `/api/v1/hello.ts` works with single router. (since now we are talking about something other than middleware, a new thread would be nice idea to not loose the discussion :)

**Reply by Shahriar-Rahi** [reply]
Yes, here I'm preferring to handle all routes in a single file because I only need these routes to proxy my secret routes. It becomes difficult to proxy huge number of endpoints if we have to create a different file for each endpoint. However, will wait for the fix. Thanks for taking the time to reply :).

**Reply by Baloche** [reply]
Yup ! We went for a custom `defineEventHandler` function that we use for the routes we want to protect : 

```ts
export function defineAuthenticatedEventHandler<T>(
  handler: (event: CompatibilityEvent, user: User) => T | Promise<T>
) {
  return defineEventHandler<T>(async (event) => {
    try {
      const user = await useAuth(event)
      return handler(event, user)
    } catch (err) {
      sendError(event, err)
    }
  })
}
```

The `useAuth` method get the user from the current jwt cookie and throw a 401 error if expired or doesn't exists.


Here is the definition of the route used to get the logged in user : 
```ts
// /server/api/me.get.ts
export default defineAuthenticatedEventHandler((_, user) => {
  return user
})

```

**Reply by pi0** [reply]
Nice wrapper method!

**Reply by mipruch** [reply]
@Baloche Hey! That looks great and I think the code might resolve my problem. Where did you define the function. In composables? 
```typescript
import jwt from "jsonwebtoken";

const secretKey = process.env.JWT_SECRET!;

export function defineAuthenticatedEventHandler<T>(handler: (event: Event) => T | Promise<T>) {
	return defineEventHandler(async (event) => {
		const usersJWT = getHeader(event, "Authorization");
		if (!usersJWT) {
			throw createError({
				statusCode: 401,
				statusMessage: "Unauthorized",
			});
		}

		const hash = usersJWT.split(" ")[1];
		jwt.verify(hash, secretKey!, function (err, decoded) {
			if (err) {
				throw createError({
					statusCode: 401,
					statusMessage: "Unauthorized",
				});
			}
		});
	});
}

``` 

And for the API route:
```typescript
import { defineAuthenticatedEventHandler } from "@/server/services/middleware";

export default defineAuthenticatedEventHandler((event) => {
	return "authorized";
});

``` 

But I keep getting this error. The Nuxt doesn't know that that is an API endpoint:
```typescript
{
    "url": "/api/test",
    "statusCode": 404,
    "statusMessage": "Cannot find any route matching /api/test.",
    "message": "Cannot find any route matching /api/test.",
    "stack": ""
}
```

**Reply by slavanossar** [reply]
@mipruch You can put them in `server/utils`, it's mentioned in the docs [here](https://nuxt.com/docs/guide/directory-structure/server#server-utilities).
